crates/hir_ty/src/infer/closure.rs

```diff
@@ -0,0 +1,95 @@
+//! Inference of closure parameter types based on the closure's expected type.
+
+use chalk_ir::{AliasTy, FnSubst, WhereClause};
+use hir_def::HasModule;
+use smallvec::SmallVec;
+
+use crate::{
+    to_chalk_trait_id, utils, ChalkTraitId, DynTy, FnPointer, FnSig, Interner, Substitution, Ty,
+    TyKind,
+};
+
+use super::{Expectation, InferenceContext};
+
+impl InferenceContext<'_> {
+    pub(super) fn deduce_closure_type_from_expectations(
+        &mut self,
+        closure_ty: &Ty,
+        sig_ty: &Ty,
+        expectation: &Expectation,
+    ) {
+        let expected_ty = match expectation.to_option(&mut self.table) {
+            Some(ty) => ty,
+            None => return,
+        };
+
+        // Deduction from where-clauses in scope, as well as fn-pointer coercion are handled here.
+        self.coerce(closure_ty, &expected_ty);
+
+        // Deduction based on the expected `dyn Fn` is done separately.
+        if let TyKind::Dyn(dyn_ty) = expected_ty.kind(&Interner) {
+            if let Some(sig) = self.deduce_sig_from_dyn_ty(dyn_ty) {
+                let expected_sig_ty = TyKind::Function(sig).intern(&Interner);
+
+                self.unify(sig_ty, &expected_sig_ty);
+            }
+        }
+    }
+
+    fn deduce_sig_from_dyn_ty(&self, dyn_ty: &DynTy) -> Option<FnPointer> {
+        // Search for a predicate like `<$self as FnX<Args>>::Output == Ret`
+
+        let fn_traits: SmallVec<[ChalkTraitId; 3]> =
+            utils::fn_traits(self.db.upcast(), self.owner.module(self.db.upcast()).krate())
+                .map(|tid| to_chalk_trait_id(tid))
+                .collect();
+
+        for bound in dyn_ty.bounds.map_ref(|b| b.iter(&Interner)) {
+            let bound = bound.map(|b| b.clone()).fuse_binders(&Interner);
```

Comment:
hmm, why the `fuse_binders`? unless I'm missing something, that'll shift the indexes of all lifetimes one up, and we'll still have the binder for the self type. I think it would make more sense to `substitute` something for the self type?

Reply:
This should bind both the self type and any lifetimes in the same binder, right? I did that so I could just put everything into the `FnPointer` without further adjustments. Though I guess `FnPointer` can only bind lifetimes, not types.

Maybe we should instead assert that the self type isn't part of the types we extract here? There isn't any correct type we can substitute here, right? (rustc uses a special dummy type for this, but that seems like a hack)

Reply:
We could just use the error type.

Reply:
> Maybe we should instead assert that the self type isn't part of the types we extract here? There isn't any correct type we can substitute here, right? (rustc uses a special dummy type for this, but that seems like a hack)

We'd still have to shift bound variables in that case (maybe there should be an "unwrap binders" operation that does this)